src/com/google/enterprise/adaptor/DocumentHandler.java

```diff
@@ -1000,8 +1001,40 @@ private void transform() {
       Map<String, String> params = new HashMap<String, String>();
       params.put("DocId", docId.getUniqueId());
       params.put("Content-Type", contentType);
+      if (null != lastModified) {
+        params.put("Last-Modified-Millis-UTC", "" + lastModified.getTime());
+      }
+      String origDisplayUrl = null;
+      if (null != displayUrl) {
+        origDisplayUrl = "" + displayUrl;
+      }
+      params.put("Display-URL", origDisplayUrl);
+      params.put("Crawl-Once", "" + crawlOnce);
+      params.put("Lock", "" + lock);
       transform.transform(metadata, params);
       contentType = params.get("Content-Type");
+      try {
+        final String lmMillisUTC = params.get("Last-Modified-Millis-UTC");
+        if (!Strings.isNullOrEmpty(lmMillisUTC)) {
+          final long lm = Long.parseLong(lmMillisUTC);
+          if (lastModified.getTime() != lm) {
+            lastModified = new Date(lm);
+          }
+        }
+      } catch (NumberFormatException e) {
+        log.log(Level.FINER,
+            "Failed changing last-modified date {0}", params.get("Last-Modified-Millis-UTC"));
+      }
+      try {
+        final String du = params.get("Display-URL");
+        if (!Strings.isNullOrEmpty(du)) {
+          displayUrl = new URI(du);
```

Comment:
At this point we know that du is not null, so we could safely:

if (!du.equals(origDisplayUrl)) {
  displayUrl = new URI(du);
}

In most cases du will equal origDisplayUrl and we can then in most cases skip all the parsing that URI does to get all the various parts it gets (host, port, path, query, etc.).

Also, consider changing origDisplayUrl to origDisplayUrlStr.